### PR TITLE
Add API endpoint to create user track recommendations

### PR DESCRIPTION
**Pull request by paramsingh**
# Problem

We want users to be able to recommend tracks from their listens to their followers.

# Solution

This PR does the following:
* Adds an API endpoint to create track recommendation events.
* Adds a barebones JS client method to use on the frontend side.

### CODE REVIEW

File: admin/sql/updates/2021-03-02-add-user-recommendation-event-table.sql [hunk 1]
```diff
@@ -1,6 +1,6 @@
 BEGIN;
 
-CREATE TYPE user_timeline_event_type_enum AS ENUM('recording');
+CREATE TYPE user_timeline_event_type_enum AS ENUM('recording_recommendation');
```
Comment (paramsingh) [inline admin/sql/updates/2021-03-02-add-user-recommendation-event-table.sql:3]: This was a change that should have been in #1309 but I missed it.

File: manage.py [hunk 2]
```diff
@@ -4,6 +4,8 @@
 from werkzeug.serving import run_simple
 import os
 import click
+import sqlalchemy
+from time import sleep
```
Comment (paramsingh) [inline manage.py:8]: Noticed some missing imports in manage.py so thought I'd add them

**Comment by paramsingh** [comment]
I'll fix the pep8 things after code review tomorrow,

File: listenbrainz/webserver/views/user_timeline_event_api.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+from listenbrainz.webserver.views.api_tools import validate_auth_header
+from listenbrainz.webserver.rate_limiter import ratelimit
+
+user_timeline_event_api_bp = Blueprint('user_timeline_event_api_bp', __name__)
+
+
+@user_timeline_event_api_bp.route('/create-user-recommendation/recording', methods=['POST', 'OPTIONS'])
+@crossdomain(headers="Authorization, Content-Type")
+@ratelimit()
+def create_user_recommendation_event():
```
Comment (mayhem) [inline listenbrainz/webserver/views/user_timeline_event_api.py:40]: The function name should also be changed to reflect the updated URL, I think. create_timeline_event_recroding ?

File: listenbrainz/webserver/views/user_timeline_event_api.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    user = validate_auth_header()
+
+    try:
+        data = ujson.loads(request.get_data())
+    except:
+        raise APIBadRequest("Invalid JSON")
+
+    try:
+        metadata = RecordingRecommendationMetadata(**data['metadata'])
+    except pydantic.ValidationError:
```
Comment (mayhem) [inline listenbrainz/webserver/views/user_timeline_event_api.py:50]: Can we pass a more useful message to the user?

Comment (mayhem) [reply listenbrainz/webserver/views/user_timeline_event_api.py:50]: In the block above as well.

File: listenbrainz/tests/integration/test_user_timeline_event_api.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    def test_it_writes_an_event_to_the_database(self):
+
+        metadata = {
+            'artist_name': 'Kanye West',
+            'track_name': 'Fade',
+            'artist_msid':  str(uuid.uuid4()),
+            'recording_msid': str(uuid.uuid4()),
+        }
+        r = self.client.post(
+            '/1/user-timeline-event/create-user-recommendation/recording',
```
Comment (mayhem) [inline listenbrainz/tests/integration/test_user_timeline_event_api.py:49]: Use url_for here?

**Review by mayhem — CHANGES_REQUESTED** [review]
See comments -- also docstrings are missing and I would like to see them at the very least for the API endpoints.

File: data/model/user_timeline_event.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import pydantic
+
+class UserTimelineEventType(Enum):
+    RECORDING_RECOMMENDATION = 'recording_recommendation'
+
+class RecordingRecommendationMetadata(pydantic.BaseModel):
+    artist_name: str
+    track_name: str
+    release_name: Optional[str]
+    spotify_id: Optional[str]
```
[the remark this answers is not in the document]
Comment (paramsingh) [reply data/model/user_timeline_event.py]: Removed the spotify ID here, I was keeping it as it'd be helpful for brainzplayer, but I think we rely on search anyways, so should be ok.

File: listenbrainz/db/user_timeline_event.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+
+def create_user_track_recommendation_event(user_id: int, metadata: RecordingRecommendationMetadata) -> UserTimelineEvent:
+    return create_user_timeline_event(
+        user_id=user_id,
+        event_type=UserTimelineEventType.RECORDING_RECOMMENDATION,
+        metadata=metadata,
+    )
+
+def get_user_timeline_events(user_id: int, event_type: UserTimelineEventType, count: int = 50) -> List[UserTimelineEvent]:
```
[the remark this answers is not in the document]
Comment (paramsingh) [reply listenbrainz/db/user_timeline_event.py:65]: These functions were only really for tests, I'm gonna add more pagination support in future PRs focussed on retrieving these events. This PR is only about creation of the events.

File: data/model/user_timeline_event.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mayhem) [reply data/model/user_timeline_event.py]: 👍

**Comment by paramsingh** [comment]
please take a look, @mayhem

**Review by mayhem — APPROVED** [review]
Thanks for making those changes! LGTM pending passing tests.

**Pull request merged**